# Week 5 Lab — Advanced Python + NumPy
**AUCA · Introduction to Big Data Analytics** · Instructor: Prince Ishimwe

**Name:** ______________________  **Student ID:** ______________  **Program:** ______________

Dataset: `week5_kigali_weather.csv` (60 days of Kigali weather, Sept–Oct 2026). Keep it next to this notebook.
**Rule of the week:** no `for` loops in Parts 3–5 (the Bonus printing loop is the only exception).

## Setup — load the data into four NumPy arrays

In [1]:
import numpy as np

FILE = "week5_kigali_weather.csv"

# columns: 0 day · 1 date · 2 temp_c · 3 rainfall_mm · 4 humidity_pct
dates = np.loadtxt(FILE, delimiter=",", skiprows=1, usecols=1, dtype=str)
temps, rain, hum = np.genfromtxt(FILE, delimiter=",", skip_header=1,
                                 usecols=(2, 3, 4), unpack=True)

print("Loaded", len(dates), "days:", dates[0], "to", dates[-1])
print("temps[:5] =", temps[:5])

Loaded 60 days: 2026-09-01 to 2026-10-30
temps[:5] = [24.4 25.9 24.5 26.  24.6]


## Part 1 · List Comprehensions (6 marks)

In [2]:
temps_list = [24.4, 25.9, 28.6, 19.3, 26.0, 22.1]   # given for Part 1
raw = ["24.4", "25.9", "28.6", "19.3"]               # given for Ex 1.3

### Exercise 1.1 — Transform (2 marks)

In [3]:
fahrenheit = [t * 9/5 + 32 for t in temps_list]
print(fahrenheit)

[75.92, 78.62, 83.48, 66.74000000000001, 78.8, 71.78]


### Exercise 1.2 — Filter (2 marks)

In [4]:
hot = [t for t in temps_list if t >= 26]
cool_rounded = [round(t) for t in temps_list if t < 24]
print("hot:", hot)
print("cool_rounded:", cool_rounded)

hot: [28.6, 26.0]
cool_rounded: [19, 22]


### Exercise 1.3 — Convert types (2 marks)

In [5]:
floats = [float(s) for s in raw]
print(floats)
print("Average:", sum(floats) / len(floats))

[24.4, 25.9, 28.6, 19.3]
Average: 24.55


## Part 2 · lambda (4 marks)

### Exercise 2.1 — Rank with `sorted` (2 marks)

In [6]:
results = [("Aline", 85), ("Eric", 78), ("Diane", 91), ("Samuel", 92), ("Grace", 76)]

ranked = sorted(results, key=lambda pair: pair[1], reverse=True)
print("Top 3:", ranked[:3])

Top 3: [('Samuel', 92), ('Diane', 91), ('Aline', 85)]


### Exercise 2.2 — `max` with a key (2 marks)

In [7]:
rain_log = [(3, 0.0), (7, 12.5), (11, 3.2), (14, 26.2), (18, 8.9)]  # (day, rainfall_mm)

wettest = max(rain_log, key=lambda pair: pair[1])
print(f"Wettest day: day {wettest[0]} with {wettest[1]} mm")

Wettest day: day 14 with 26.2 mm


## Part 3 · NumPy Arrays (8 marks)

### Exercise 3.1 — The speed test (2 marks)

In [8]:
import time

N = 1_000_000
py_list = list(range(N))
np_arr = np.arange(N)

t0 = time.perf_counter()
doubled_list = []                            # pure Python loop
for x in py_list:
    doubled_list.append(x * 2)
py_time = time.perf_counter() - t0

t0 = time.perf_counter()
doubled_arr = np_arr * 2                     # vectorized NumPy
np_time = time.perf_counter() - t0

print(f"Python loop: {py_time:.4f} s")
print(f"NumPy      : {np_time:.4f} s")
print(f"NumPy was about {py_time / np_time:.0f}x faster")

# NumPy was roughly 10x faster in my run (update this to the number YOUR machine prints; it varies;
# see the number printed above). It runs the loop in optimized C instead of Python.

Python loop: 0.0834 s
NumPy      : 0.0108 s
NumPy was about 8x faster


### Exercise 3.2 — Describe the temperatures (3 marks)

In [9]:
print(f"Mean: {temps.mean():.2f} °C")
print(f"Max : {temps.max():.1f} °C")
print(f"Min : {temps.min():.1f} °C")
print(f"Std : {temps.std():.2f} °C")
# Checkpoint: mean should be 24.00 °C

Mean: 24.00 °C


Max : 28.6 °C
Min : 19.3 °C
Std : 2.61 °C


### Exercise 3.3 — The hottest day (3 marks)

In [10]:
pos = np.argmax(temps)
print(f"Hottest day: {dates[pos]} at {temps[pos]} °C")

Hottest day: 2026-09-17 at 28.6 °C


## Part 4 · Boolean Masking (10 marks)

### Exercise 4.1 — Hot days (3 marks)

In [11]:
hot_mask = temps >= 26

print(f"(a) Number of hot days: {hot_mask.sum()}")
print(f"(b) Average temperature on hot days: {temps[hot_mask].mean():.2f} °C")

(a) Number of hot days: 18
(b) Average temperature on hot days: 27.11 °C


### Exercise 4.2 — Rain report (3 marks)

In [12]:
rainy_mask = rain > 0

print(f"(a) Rainy days: {rainy_mask.sum()}")
print(f"(b) Total rainfall over 60 days: {rain.sum():.1f} mm")
print(f"(c) Heaviest single-day rainfall: {rain.max():.1f} mm")

(a) Rainy days: 16
(b) Total rainfall over 60 days: 331.2 mm
(c) Heaviest single-day rainfall: 26.2 mm


### Exercise 4.3 — Cross-array analysis (4 marks)

In [13]:
# (a) mask from rain, data from temps
print(f"(a) Avg temp on rainy days: {temps[rainy_mask].mean():.2f} °C")
print(f"    Avg temp on dry days  : {temps[~rainy_mask].mean():.2f} °C")

# (b) hot AND rainy  (parentheses around each condition)
print(f"(b) Hot AND rainy days: {((temps >= 26) & (rain > 0)).sum()}")

# (c) humidity >= 70
print(f"(c) Days with humidity >= 70%: {(hum >= 70).sum()}")

(a) Avg temp on rainy days: 24.14 °C


    Avg temp on dry days  : 23.95 °C
(b) Hot AND rainy days: 5
(c) Days with humidity >= 70%: 16


## Part 5 · Anomalies & Trends (7 marks)

### Exercise 5.1 — Anomaly detection (4 marks)

In [14]:
anomaly = temps - temps.mean()

warm_mask = anomaly > 3
print("(a) Days more than 3 °C above the mean:")
print(dates[warm_mask])

print(f"(b) Days more than 3 °C below the mean: {(anomaly < -3).sum()}")

(a) Days more than 3 °C above the mean:
['2026-09-07' '2026-09-09' '2026-09-11' '2026-09-12' '2026-09-14'
 '2026-09-16' '2026-09-17' '2026-09-19' '2026-09-22']
(b) Days more than 3 °C below the mean: 8


### Exercise 5.2 — Is the season changing? (3 marks)

In [15]:
first30 = temps[:30].mean()
last30 = temps[30:].mean()
print(f"First 30 days average: {first30:.2f} °C")
print(f"Last 30 days average : {last30:.2f} °C")
print(f"Change: {last30 - first30:+.2f} °C")

# Conclusion: Kigali got COOLER. The last 30 days average about 4.4 °C lower
# than the first 30, so the season cooled as September turned into October.

First 30 days average: 26.21 °C
Last 30 days average : 21.78 °C
Change: -4.43 °C


## Bonus (+3) — Top 5 wettest days

In [16]:
pairs = sorted(zip(dates, rain), key=lambda p: p[1], reverse=True)
for d, mm in pairs[:5]:          # the only loop allowed: printing the bonus
    print(f"{d}: {mm} mm")

2026-09-07: 26.2 mm
2026-09-14: 25.1 mm
2026-09-21: 24.0 mm
2026-09-05: 22.9 mm
2026-09-28: 22.9 mm


## Part 6 · Reflection

**1. What did NumPy hide from you — and why was it still important to learn the loop version first?**

`temps.mean()` hides the whole mechanism: creating a running total, walking through every element one at a time, counting them, and dividing at the end. It also hides that the work is done by fast compiled C code rather than by Python's slower loop, which is why the speed test showed such a large gap. Learning the loop version first matters because I now understand *what* `mean()` is actually computing, so I can trust it, debug it when something looks wrong (for example a missing value or wrong data type), and write my own logic when no built-in method exists. Without the loop version, `mean()` would just be a magic word; with it, I know what the magic does.

**2. One insight about Kigali's weather**

The season clearly cooled and got more humid: the first 30 days averaged about 26.2 °C, while the last 30 averaged about 21.8 °C, and average humidity rose from about 58% to 66%. Rain fell on only 16 of 60 days, yet those days delivered 331 mm in total, and the five wettest days were spread across the period (7, 14, 21 and 28 Sept, plus 5 Sept) rather than bunched together. Rainy and dry days had almost the same average temperature (about 24.1 °C vs 23.9 °C), so rain by itself did not explain the cooling. The drop in temperature looks like a seasonal shift rather than an effect of individual rainy days, and the wettest days fell in the warmer first half (183 mm in the first 30 days vs 148 mm in the last 30).